site/source/docs/api_reference/val.h.rst

```diff
@@ -48,7 +46,13 @@ Guide material for this class can be found in :ref:`embind-val-guide`.
     }
 
   See :ref:`embind-val-guide` for other examples.
+  
+
+  .. warning:: JavaScript values aren't shared across threads, so neither can ``val``. If you want to create a ``val`` binding to some JavaScript value and access it from multiple threads, you must create a unique binding per each thread using e.g. ``thread_local``:
```

Comment:
```suggestion
  .. warning:: JavaScript values aren't shared across threads, so neither are ``val`s`. If you want to create a ``val`` binding to some JavaScript value and access it from multiple threads, you must create a unique binding per each thread using e.g. ``thread_local``:
```

Reply:
Perhaps this should stress more that the value is different in each thread? The current text suggests "access it from multiple threads" which could be misunderstood.

Reply:
```suggestion
  .. warning:: JavaScript values can't be shared across threads, so neither can ``val``. For example, if you want to cache some JavaScript global as a ``val``, you need to retrieve that global separately in each thread and bind to per-thread ``val`` bindings with ``thread_local``:
```

Agreed. How about this?

Reply:
Hmm, better, but I'm still not sure what this is saying. Maybe just me?

I'd only write something like
```
warning:: JavaScript values can't be shared across threads, so neither can val objects (as they refer to JavaScript values).
```

Reply:
I wonder if we can/should we make the `val` class itself `thread_local` since it make not sense to be anything else?

Reply:
> I'd only write something like

That seems to be almost exactly what I had in my last suggestion? I still think it's worth following up with an actual example of correct usage for common use-case if that's what you want to remove.

Reply:
> Oh wait no, we can't make classes or members as thread local IIUC. Sorry for the noise.

Yeah that could've been nice. The only thing we could do to help is add thread id to each `val` and assert it in a debug mode.

Reply:
I kind agree with @kripken that mentioning `thread-local` might be a little misleading, since it implies that folks should use C++ `thread_local` with `val`.  I would have thought that in most cases folks probably get by with just access the val from a single thread (e.g. by running all their JS-interaction code on just one thread).

Or are you actually suggesting that when folks write threaded code they decorate their `val` variables with `thread_local`?

Reply:
In other words maybe "not shareable between threads" works better here than "thread-local" ... even though they are basically the same thing.

Obviously this is not a huge deal..

Reply:
Probably should explain the motivation: in general caching is useful because `val` access is not cheap, and involves string decoding, runtime property access etc. etc. which can add up in hot paths. This is something to be solved in the new bindings system, but meanwhile I believe it's worth at least mentioning how to correctly cache variables to reduce those costs.

For example, implementation of `val`'s own `val::hasOwnProperty`: https://github.com/emscripten-core/emscripten/blob/e274e49bfb2022313f5697dc5560c19f5f13d81a/system/include/emscripten/val.h#L401-L403

With this suggestion could be rewritten as:

```cpp
bool hasOwnProperty(const char* key) const {
    thread_local hasOwnProperty = val::global("Object")["prototype"]["hasOwnProperty"];
    return hasOwnProperty.call<bool>("call", *this, val(key));
}
```

So that all those strings and the method are retrieved only once per thread, and then reused on each next call.

Reply:
Great! I've applied latest revised warning text from the comment thread.

Reply:
> That seems to be almost exactly what I had in my last suggestion? 

Yes, exactly - I was trying to keep the text that I understood, and remove the part about a `thread_local` variable which I either don't understand, or which seems confusing.